### PR TITLE
Replace TTK for PSRule

### PR DESCRIPTION
**Pull request by VeraBE**
Replace [TTK](https://github.com/Azure/arm-ttk) for [PSRule](https://github.com/microsoft/PSRule) as the core of the PowerShell rule engine

I'm not sure about still having an option like `run-powershell`, I think we should always run all the rules, thoughts?

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] The pull request does not introduce breaking changes.

### [General Guidelines](../CONTRIBUTING.md)
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request is clear and informative.
- [x] I have added myself to the 'assignees'.
- [x] I have added 'linked issues' if relevant.

### [Testing Guidelines](../CONTRIBUTING.md#tests)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by BernieWhite — COMMENTED** [review]
@VeraBE Nice work. I tried running it locally and got some results.

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs [hunk 1]
```diff
@@ -102,108 +44,60 @@ public IEnumerable<IEvaluation> AnalyzeTemplate(TemplateContext templateContext)
...
-            this.logger?.LogError(summary);
-            foreach (var error in errors)
-            {
-                this.logger?.LogError(error.ErrorDetails.Message);
-            }
+                    Format = InputFormat.Json
+                },
+                Output = outputOption
+            };
+            var resources = templateContext.ExpandedTemplate.InsensitiveToken("resources");
```
[the remark this answers is not in the document]
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs:93]: Process(...) expects a PSObject, so I tried your suggestion changing Values<JObject> for Values<PSObject> and it does't seem to work, I only get the results of the file analysis :/ I tried passing the resources as strings and other variations but no luck yet, any ideas?

**Comment by BernieWhite** [comment]
@VeraBE Added update to merge into this PR via https://github.com/Azure/template-analyzer/pull/273

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs [hunk 2]
```diff
@@ -67,17 +56,13 @@ public void AnalyzeTemplate_NotUsingPowerShell_NoPowerShellViolations()
...
+            var evaluations = templateAnalyzerWithoutPowerShell.AnalyzeTemplate(template);
 
             // There should be no PowerShell rule evaluations because the PowerShell engine should not have run
             Assert.IsFalse(evaluations.Any(e => e is PowerShellRuleEvaluation));
 
             // Analyze with PowerShell enabled
-            evaluations = templateAnalyzer.AnalyzeTemplate(
-                template,
-                templateFilePath: @"..\..\..\..\Analyzer.PowerShellRuleEngine.UnitTests\templates\error_without_line_number.json"); // This file has violations from PowerShell rules
+            evaluations = templateAnalyzer.AnalyzeTemplate(template, templateFilePath: "aTemplateFilePath");
```
Comment (anaismiller) [inline src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:65]: Seems the function name is wrong

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs [hunk 3]
```diff
@@ -22,14 +24,23 @@ public static void AssemblyInitialize(TestContext context)
         }
 
         [DataTestMethod]
-        [DataRow("success.json", 0, new int[] { }, DisplayName = "Base template")]
-        [DataRow("error_without_line_number.json", 1, new int[] { 0 }, DisplayName = "Template with an error reported without a line number")]
-        [DataRow("error_with_line_number.json", 1, new int[] { 9 }, DisplayName = "Template with an error reported with a line number")]
-        [DataRow("warning.json", 1, new int[] { 0 }, DisplayName = "Template with a warning")]
-        [DataRow("repeated_error_same_message_same_lines.json", 1, new int[] { 0 }, DisplayName = "Template with an error found in multiple places, reported with the same message and line number for each")]
+        [DataRow("template_and_resource_level_results.json", 12, new int[] { 1, 1, 1, 1, 8, 14, 17, 1, 17, 17, 1, 17 }, DisplayName = "Template with errors reported in both analysis stages")]
+        // TODO add test case for error, warning (rule with severity level of warning?) and informational (also rule with that severity level?)
```
Comment (anaismiller) [inline src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:28]: Why did you remove them?

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs [hunk 4]
```diff
@@ -102,108 +45,51 @@ public IEnumerable<IEvaluation> AnalyzeTemplate(TemplateContext templateContext)
...
+                throw new ArgumentException($"{nameof(TemplateContext.ResourceMappings)} must not be null.", nameof(templateContext));
             }
 
-            return evaluations;
-        }
-
-        private void PreProcessErrors(dynamic error, Dictionary<string, SortedSet<int>> uniqueErrors)
-        {
-            var lineNumber = 0;
+            // TODO: Temporary work-around: write template to disk so PSRule will analyze it
```
Comment (anaismiller) [inline src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs:58]: Don't have context - why is this work-around needed/what is the future plan?

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEvaluation.cs [hunk 5]
```diff
@@ -26,7 +29,7 @@ public class PowerShellRuleEvaluation : IEvaluation
         public string HelpUri { get; }
 
         /// <inheritdoc/>
-        public Severity Severity { get; } = Severity.Medium;
```
Comment (anaismiller) [inline src/Analyzer.PowerShellRuleEngine/PowerShellRuleEvaluation.cs:29]: This was because the default is medium and is not a necessary parameter

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 3 above)
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:28]: Because those test cases were covering BPA code paths that no longer exist when using PSRule

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs (hunk 2 above)
Comment (VeraBE) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs:65]: Why do you think that?

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEvaluation.cs (hunk 5 above)
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine/PowerShellRuleEvaluation.cs:29]: Is this a question? That was there because TTK didn't have severities so we were using Medium for everything, now the severity is always set in the constructor

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs (hunk 4 above)
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs:58]: Because PSRule needs to read the template from a file to do some of the checks. The future plan is to make the needed fixes in there so that we can just pass the template string

File: src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs (hunk 2 above)
Comment (anaismiller) [reply src/Analyzer.Core.UnitTests/TemplateAnalyzerTests.cs]: Whoops, my error.

**Review by anaismiller — APPROVED** [review]

**Review by BernieWhite — APPROVED** [review]
@VeraBE Looks great. Nice work.

File: src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs (hunk 3 above)
Comment (BernieWhite) [reply src/Analyzer.PowerShellRuleEngine.UnitTests/PowerShellRuleEngineTests.cs:28]: @VeraBE TODO comments true. Sorry this one skipped my memory.

We can:

- I need to update some rules in PSRule for Azure with different error levels, so we could look at adding these. or
- Log an issue assigned to me to create a custom rule for testing in this repository.

**Review by nonik0 — CHANGES_REQUESTED** [review]

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs [hunk 6]
```diff
@@ -102,108 +45,51 @@ public IEnumerable<IEvaluation> AnalyzeTemplate(TemplateContext templateContext)
...
-                {
-                    LogPowerShellErrors(powershell.Streams.Error, $"There was an error unblocking scripts in path '{directory}'.");
-                }
+            pipeline.End();
 
-                powershell.Commands.Clear();
-            }
-        }
+            // Remove temporary file:
+            File.Delete(templateFile);
```
[the remark this answers is not in the document]
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs:90]: I'm tempted of leaving it like this because is temporary code, and the temp files will auto delete anyways

[the remark this answers is not in the document]
Comment (anaismiller) [reply src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs]: I agree with Nick about a try/finally

File: README.md [hunk 7]
```diff
@@ -35,6 +35,7 @@ Argument | Description
 **(Optional)** `--report-format` | Valid formats:<br/>*Console*: output results to the console in plain text. **(default)**<br/>*Sarif*: output results to a file in [SARIF](https://sarifweb.azurewebsites.net) format.
 `-o` or `--output-file-path` | **(Required if `--report-format` is *Sarif*)**  File path to output SARIF results to.
 **(Optional)** `-v` or `--verbose` | Shows details about the analysis
+**(Optional)** `-run-powershell` | Also runs the PowerShell-based rules
```
[the remark this answers is not in the document]
Comment (anaismiller) [reply README.md]: Wait, I agree with Nick that it should be `--` to follow the pattern above. `-` is used for the single char shortcut.

[the remark this answers is not in the document]
Comment (VeraBE) [reply README.md]: I already made the change

**Review by nonik0 — APPROVED** [review]

File: src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs [hunk 8]
```diff
@@ -102,108 +45,51 @@ public IEnumerable<IEvaluation> AnalyzeTemplate(TemplateContext templateContext)
...
-        }
-
-        private void PreProcessErrors(dynamic error, Dictionary<string, SortedSet<int>> uniqueErrors)
-        {
-            var lineNumber = 0;
+            // TODO: Temporary work-around: write template to disk so PSRule will analyze it
+            var tempTemplateFile = Path.GetTempFileName();
+            File.WriteAllText(tempTemplateFile, templateContext.ExpandedTemplate.ToString());
+            var templateFile = tempTemplateFile.Replace(".tmp", ".json");
+            File.Move(tempTemplateFile, templateFile);
```
[the remark this answers is not in the document]
Comment (VeraBE) [reply src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs]: But I made the change: https://github.com/Azure/template-analyzer/pull/269/commits/bc2d420ec4a392bf8b5f42aa11a41a692fea5002
